internal/configs/resource.go

```diff
@@ -731,6 +741,25 @@ func decodeProviderConfigRef(expr hcl.Expression, argName string) (*ProviderConf
 		ret.AliasRange = aliasStep.SourceRange().Ptr()
 	}
 
+	if len(traversal) > 2 {
+		indexStep, ok := traversal[2].(hcl.TraverseIndex)
+		if !ok || ret.KeyExpression != nil {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid provider configuration reference",
+				Detail:   "Provider name must either stand alone or be followed by a period and then a configuration alias.",
+				Subject:  traversal[2].SourceRange().Ptr(),
+			})
+			return ret, diags
+		}
+
+		ret.KeyExpression = hcl.StaticExpr(indexStep.Key, traversal.SourceRange())
+	}
+
+	if len(ret.Alias) == 0 && ret.KeyExpression != nil {
+		panic("Alias must be set when using a keyed expression! (todo better error message)")
```

Comment:
I think this is a "should never happen case", right? It seems we already have logic in the main decode function that rejects a configuration which has `for_each` without `alias`, so I think a panic is fine here.

Reply:
[fc15d70](https://github.com/opentofu/opentofu/pull/2105/commits/fc15d70c561880f92f2afc36baae46a3a1858a7d)

Reply:
It's actually a valid error case and has been updated with better diagnostics